packages/ui/src/views/user-management/components/dialogs/hooks/use-dialog-data.ts

```diff
@@ -0,0 +1,40 @@
+import { generateAlphaNumericHash } from '@/utils/utils'
+import { DialogLabels } from '@/views/user-management/components/dialogs'
+import { useDialogs } from '@/views/user-management/providers/dialogs-provider'
+import { useUserManagementStore } from '@/views/user-management/providers/store-provider'
+import { UsersProps } from '@/views/user-management/types'
+
+export const useDialogData = () => {
+  const { useAdminListUsersStore } = useUserManagementStore()
+
+  const { setUser, setPassword, setGeteneratePassword } = useAdminListUsersStore()
+
+  const { openDialog } = useDialogs()
+
+  const prepareDialogData = (user: UsersProps | null, dialogType: DialogLabels) => {
+    if (user) setUser(user)
+
+    if (dialogType === DialogLabels.RESET_PASSWORD) {
+      setGeteneratePassword(false)
+      setPassword(generateAlphaNumericHash(10))
+
+      return
+    }
+
+    if (dialogType === DialogLabels.CREATE_USER) {
+      setGeteneratePassword(true)
+      setPassword(generateAlphaNumericHash(10))
+
+      return
+    }
+  }
+
+  const handleDialogOpen = (user: UsersProps | null, dialogType: DialogLabels) => {
+    prepareDialogData(user, dialogType)
+    openDialog(dialogType)
+  }
+
+  return {
+    handleDialogOpen
+  }
```

Comment:
Worth pointing out that this will not be dependency array safe. Easiest way is to wrap this whole bit in a `useMemo`
```suggestion
  return useMemo(() => {
    const prepareDialogData = (user: UsersProps | null, dialogType: DialogLabels) => {
      if (user) setUser(user)

      if (dialogType === DialogLabels.RESET_PASSWORD) {
        setGeteneratePassword(false)
        setPassword(generateAlphaNumericHash(10))

        return
      }

      if (dialogType === DialogLabels.CREATE_USER) {
        setGeteneratePassword(true)
        setPassword(generateAlphaNumericHash(10))

        return
      }
    }

    const handleDialogOpen = (user: UsersProps | null, dialogType: DialogLabels) => {
      prepareDialogData(user, dialogType)
      openDialog(dialogType)
    }

    return {
      handleDialogOpen
    }
  }, [setUser, setPassword, setGeteneratePassword, openDialog])
```

Reply:
Thanks for your comment! I’ve checked the current usage of `handleDialogOpen` and tried to evaluate the benefits of memoization. From what I see, it’s mostly used in static components that don’t expect frequent re-renders and aren’t memoized (e.g., `NoData` and others). As a result, these components will re-render on every render anyway, regardless of whether handleDialogOpen is stable or not. Am I missing something here? just want to clarify this for my own understanding. Thanks! :raised_hands:

Reply:
The two most important things about memoization are to cut down on needless expensive processing (not applicable here) and to stabilise references. Since objects, arrays and functions are always passed by reference, if they are created in the flow of a react function component (as in the case of hooks), then they will be redefined on every render.

Often that doesn't really matter; it's inefficient, but ¯\\_(ツ)_/¯. However, if that function, object or array is used in the dependency array of something like a `useMemo` or `useEffect`, then it will trigger on _every_ render. So when passing functions, objects and arrays around, it's a good practice to stabilise their reference. For objects and arrays, `useMemo` is great for that and for functions, `useCallback` is ideal. In this case you're passing back an object with a single function, so you could go either way.

To demonstrate the use-case, the example below will trigger the `useEffect` on _every_ render.

```typescript jsx
function useMyAwesomeThing () {
  const getNumber = () => Math.random() * 1000

  return { getNumber }
}

function MyComponent () {
  const { getNumber } = useMyAwesomeThing()
  
  useEffect({
    alert(getNumber())
  }, [getNumber])

  return <p>Hello</p>
}
```

You can leave `getNumber` out of the dependency array, but ESlint will complain that you left it out and you'll have to disable the lint rule to get around it. And _most importantly_, it assumes you know exactly where and how that hook will be used throughout the codebase and that no one will ever use it in a dependency array.

You can fix the above by stabilising the reference. You can either stabilise the object you're returning...
```typescript jsx
function useMyAwesomeThing () {
  return useMemo(() => {
    const getNumber = () => Math.random() * 1000

    return { getNumber }
  }, [])
}
```
or the function.
```typescript jsx
function useMyAwesomeThing () {
  const getNumber = useCallback(() => Math.random() * 1000, [])

  return { getNumber }
}
```

Either of the above will make sure that `getNumber` is dependency array safe on every render.

In your case, this might not be an issue in the slightest. But it's a good habit to assume that you have no control over where your component/hook etc will be used and plan for the worst. Generally you'll hit less bugs and cut down on needless re-renders by being mindful of when you're creating and passing new functions, objects and arrays on each render.


I made up a simple CodePen a couple of years back to demonstrate. To use it, just uncomment the commented out lines one-by-one and you'll see the effect on the renders https://codepen.io/knagurski/pen/VwVWPQJ

Reply:
Thanks @knagurski  for the detailed explanation and examples :heart: